pkg/kubelet/kubelet.go

```diff
@@ -2813,6 +2824,11 @@ func (kl *Kubelet) setNodeStatusMachineInfo(node *api.Node) {
 				"Node %s has been rebooted, boot id: %s", kl.nodeName, info.BootID)
 		}
 		node.Status.NodeInfo.BootID = info.BootID
+
+		// If no cloud provider is defined - use the one detected by cadvisor
+		if kl.cloud == nil {
+			kl.updateCloudProviderFromMachineInfo(node, info)
```

Comment:
I'm not excited about updating the `node.Spec` in a `setNodeStatus***` method.  In fact, it won't function properly when called through this chain: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L3030.  Can you perform these checks here where the spec is modified?  https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L969